File: src/main/java/raylras/zen/code/type/ListType.java
package raylras.zen.code.type;

import raylras.zen.code.CompilationEnvironment;
import raylras.zen.code.SymbolProvider;
import raylras.zen.code.symbol.Operator;
import raylras.zen.code.symbol.Symbol;
import raylras.zen.code.symbol.SymbolFactory;
import raylras.zen.code.symbol.SymbolGroup;

import java.util.Objects;
import java.util.function.UnaryOperator;

public class ListType extends Type implements SymbolProvider {

    private final Type elementType;

    public ListType(Type elementType) {
        this.elementType = elementType;
    }

    public Type getElementType() {
        return elementType;
    }

    @Override
    public SymbolGroup getSymbols() {
        return SymbolFactory.builtinSymbols()
                .variable("length", IntType.INSTANCE, Symbol.Modifier.VAL)
                .function("remove", VoidType.INSTANCE, params -> params.parameter("index", IntType.INSTANCE))
                .operator(Operator.INDEX_GET, elementType, params -> params.parameter("index", IntType.INSTANCE))
                .operator(Operator.INDEX_SET, elementType, params ->
                        params.parameter("index", IntType.INSTANCE).parameter("element", elementType)
                )
                .operator(Operator.ADD, this, params -> params.parameter("element", elementType))
                .operator(Operator.ITERATOR, this, UnaryOperator.identity())
                .build();
    }

    @Override
    public boolean isInheritedFrom(Type type) {
        if (type instanceof ListType that && elementType.isInheritedFrom(that.getElementType())) {
            return true;
        }
        return super.isInheritedFrom(type);
    }

    @Override
    public boolean isCastableTo(Type type, CompilationEnvironment env) {
        if (type instanceof ListType that && elementType.isCastableTo(that.getElementType(), env)) {
            return true;
        }
        if (type instanceof ArrayType that && elementType.isAssignableTo(that.getElementType(), env)) {
            return true;
        }
        return super.isCastableTo(type, env);
    }

    @Override
    public String toString() {
        return "[" + elementType + "]";
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ListType symbols = (ListType) o;
        return Objects.equals(elementType, symbols.elementType);
    }

    @Override
    public int hashCode() {
        return Objects.hash(toString());
    }
}

File: core/src/main/kotlin/raylras/intellizen/type/ListType.kt
package raylras.intellizen.type

import raylras.intellizen.CompilationEnvironment
import raylras.intellizen.symbol.Modifiable.Modifier
import raylras.intellizen.symbol.Operator
import raylras.intellizen.symbol.Symbol
import raylras.intellizen.symbol.SymbolProvider
import raylras.intellizen.symbol.symbolSequence

data class ListType(val elementType: Type) : Type, SymbolProvider {
    override val typeName: String by lazy { "[${elementType.typeName}]" }

    override val simpleTypeName by lazy { "[${elementType.simpleTypeName}]" }

    override fun isCastableTo(that: Type?, env: CompilationEnvironment): Boolean {
        return when (that) {
            is ListType -> {
                elementType.isCastableTo(that.elementType, env)
            }

            is ArrayType -> {
                elementType.isCastableTo(that.elementType, env)
            }

            else -> super.isCastableTo(that, env)
        }
    }

    override fun getSymbols(env: CompilationEnvironment?): Sequence<Symbol> {
        return symbolSequence {
            variable("length", IntType, Modifier.IMPLICIT_VAL)
            function("remove", VoidType) {
                parameter("index", IntType)
            }
            operator(Operator.INDEX_GET, elementType) {
                parameter("index", IntType)
            }
            operator(Operator.INDEX_SET, elementType) {
                parameter("index", IntType)
                parameter("element", elementType)
            }
            operator(Operator.ADD, this@ListType) {
                parameter("element", elementType)
            }
            operator(Operator.FOR_IN, this@ListType)
        }
    }

    override fun toString() = typeName
}
